clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -506,12 +572,55 @@ void CIRGenFunction::buildStoreOfScalar(mlir::Value value, LValue lvalue,
 /// method emits the address of the lvalue, then loads the result as an rvalue,
 /// returning the rvalue.
 RValue CIRGenFunction::buildLoadOfLValue(LValue LV, SourceLocation Loc) {
-  assert(LV.isSimple() && "not implemented");
   assert(!LV.getType()->isFunctionType());
   assert(!(LV.getType()->isConstantMatrixType()) && "not implemented");
 
-  // Everything needs a load.
-  return RValue::get(buildLoadOfScalar(LV, Loc));
+  if (LV.isBitField())
+    return buildLoadOfBitfieldLValue(LV, Loc);
+
+  if (LV.isSimple())
+    return RValue::get(buildLoadOfScalar(LV, Loc));
+  llvm_unreachable("NYI");
+}
+
+RValue CIRGenFunction::buildLoadOfBitfieldLValue(LValue LV,
+                                                 SourceLocation Loc) {
+  const CIRGenBitFieldInfo &Info = LV.getBitFieldInfo();
+
+  // Get the output type.
+  mlir::Type ResLTy = convertType(LV.getType());
+  Address Ptr = LV.getBitFieldAddress();
+  mlir::Value Val = builder.createLoad(getLoc(Loc), Ptr);
+  auto ValWidth = Val.getType().cast<IntType>().getWidth();
+
+  bool UseVolatile = LV.isVolatileQualified() &&
+                     Info.VolatileStorageSize != 0 && isAAPCS(CGM.getTarget());
+  const unsigned Offset = UseVolatile ? Info.VolatileOffset : Info.Offset;
+  const unsigned StorageSize =
+      UseVolatile ? Info.VolatileStorageSize : Info.StorageSize;
+
+  if (Info.IsSigned) {
+    assert(static_cast<unsigned>(Offset + Info.Size) <= StorageSize);
+
+    mlir::Type typ = builder.getSIntNTy(ValWidth);
+    Val = builder.createIntCast(Val, typ);
+
+    unsigned HighBits = StorageSize - Offset - Info.Size;
+    if (HighBits)
+      Val = builder.createShiftLeft(Val, HighBits);
+    if (Offset + HighBits)
+      Val = builder.createShiftRight(Val, Offset + HighBits);
+  } else {
+    if (Offset)
+      Val = builder.createShiftRight(Val, Offset);
+
+    if (static_cast<unsigned>(Offset) + Info.Size < StorageSize)
+      Val = builder.createAnd(Val,
+                              llvm::APInt::getLowBitsSet(ValWidth, Info.Size));
+  }
+  Val = builder.createIntCast(Val, ResLTy);
+  // EmitScalarRangeCheck(Val, LV.getType(), Loc); //FIXME: TODO
```

Comment:
We're usually add a bit more description and avoid leaving commented code like this. If this is something we cannot assert on, you should create an `assert(!UnimplementedFeature:...` for it!

Reply:
done!